File: app/src/main/java/com/softartdev/notecrypt/ui/settings/security/SecurityActivity.java
package com.softartdev.notecrypt.ui.settings.security;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.NavUtils;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.Switch;

import com.softartdev.notecrypt.R;
import com.softartdev.notecrypt.ui.base.BaseActivity;

import javax.inject.Inject;

public class SecurityActivity extends BaseActivity implements SecurityView, View.OnClickListener, CompoundButton.OnCheckedChangeListener {
    @Inject
    SecurityPresenter mPresenter;

    Switch enableEncryptionSwitch;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_security);
        activityComponent().inject(this);
        mPresenter.attachView(this);
        enableEncryptionSwitch = findViewById(R.id.enable_encryption_switch);
        enableEncryptionSwitch.setChecked(mPresenter.isEncryption());
        enableEncryptionSwitch.setOnCheckedChangeListener(this);
        Button passButton = findViewById(R.id.set_password_button);
        passButton.setOnClickListener(this);
    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        if (buttonView.getId() == R.id.enable_encryption_switch) {
            mPresenter.changeEncryption(isChecked);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.set_password_button:
                mPresenter.changePassword();
                break;
        }
    }

    @Override
    public void showEncryptEnable(boolean encryption) {
        enableEncryptionSwitch.setOnCheckedChangeListener(null);
        enableEncryptionSwitch.setChecked(encryption);
        enableEncryptionSwitch.setOnCheckedChangeListener(this);
    }

    @Override
    public void showPasswordDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        LayoutInflater inflater = getLayoutInflater();
        @SuppressLint("InflateParams") View view = inflater.inflate(R.layout.dialog_password, null);
        final TextInputLayout enterPassTextInputLayout = view.findViewById(R.id.enter_password_text_input_layout);
        final TextInputEditText enterPassEditText = view.findViewById(R.id.enter_password_edit_text);
        builder.setView(view)
                .setPositiveButton(android.R.string.ok, null)
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> dialog.cancel());
        final AlertDialog alertDialog = builder.create();
        alertDialog.setOnShowListener(dialog -> {
            Button okButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE);
            okButton.setOnClickListener(v -> {
                DialogDirector pass = new PassMediator(enterPassTextInputLayout, enterPassEditText);
                if (mPresenter.enterPassCorrect(pass)) {
                    alertDialog.dismiss();
                }
            });
        });
        alertDialog.show();
    }

    @Override
    public void showSetPasswordDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        LayoutInflater inflater = getLayoutInflater();
        @SuppressLint("InflateParams") View view = inflater.inflate(R.layout.dialog_set_password, null);
        final TextInputLayout setPassTextInputLayout = view.findViewById(R.id.set_password_text_input_layout);
        final TextInputEditText setPassEditText = view.findViewById(R.id.set_password_edit_text);
        final TextInputLayout repeatSetPassTextInputLayout = view.findViewById(R.id.repeat_set_password_text_input_layout);
        final TextInputEditText repeatSetPassEditText = view.findViewById(R.id.repeat_set_password_edit_text);
        builder.setView(view)
                .setPositiveButton(android.R.string.ok, null)
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> dialog.cancel());
        final AlertDialog alertDialog = builder.create(); //TODO: try to make it global for move positive button listener above
        alertDialog.setOnShowListener(dialog -> {
            Button okButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE);
            okButton.setOnClickListener(v -> {
                DialogDirector pass = new PassMediator(setPassTextInputLayout, setPassEditText);
                DialogDirector repeatPass = new PassMediator(repeatSetPassTextInputLayout, repeatSetPassEditText);
                if (mPresenter.setPassCorrect(pass, repeatPass)) {
                    alertDialog.dismiss();
                }
            });
        });
        alertDialog.show();
    }

    @Override
    public void showChangePasswordDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        LayoutInflater inflater = getLayoutInflater();
        @SuppressLint("InflateParams") View view = inflater.inflate(R.layout.dialog_change_password, null);
        final TextInputLayout oldPassTextInputLayout = view.findViewById(R.id.old_password_text_input_layout);
        final TextInputEditText oldPassEditText = view.findViewById(R.id.old_password_edit_text);
        final TextInputLayout newPassTextInputLayout = view.findViewById(R.id.new_password_text_input_layout);
        final TextInputEditText newPassEditText = view.findViewById(R.id.new_password_edit_text);
        final TextInputLayout repeatPassTextInputLayout = view.findViewById(R.id.repeat_new_password_text_input_layout);
        final TextInputEditText repeatNewPassEditText = view.findViewById(R.id.repeat_new_password_edit_text);
        builder.setView(view)
                .setPositiveButton(android.R.string.ok, null)
                .setNegativeButton(android.R.string.cancel, (dialog, which) -> dialog.cancel());
        final AlertDialog alertDialog = builder.create();
        alertDialog.setOnShowListener(dialog -> {
            Button okButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE);
            okButton.setOnClickListener(v -> {
                DialogDirector oldPass = new PassMediator(oldPassTextInputLayout, oldPassEditText);
                DialogDirector newPass = new PassMediator(newPassTextInputLayout, newPassEditText);
                DialogDirector repeatNewPass = new PassMediator(repeatPassTextInputLayout, repeatNewPassEditText);
                if (mPresenter.changePassCorrect(oldPass, newPass, repeatNewPass)) {
                    alertDialog.dismiss();
                }
            });
        });
        alertDialog.show();
    }

    private class PassMediator implements DialogDirector {
        TextInputLayout mTextInputLayout;
        EditText mEditText;

        PassMediator(TextInputLayout textInputLayout, EditText editText) {
            mTextInputLayout = textInputLayout;
            mEditText = editText;
        }

        @Override
        public String getTextString() {
            return mEditText.getText().toString();
        }

        @Override
        public void showIncorrectPasswordError() {
            mTextInputLayout.setError(getString(R.string.incorrect_password));
        }

        @Override
        public void showEmptyPasswordError() {
            mTextInputLayout.setError(getString(R.string.empty_password));
        }

        @Override
        public void showPasswordsNoMatchError() {
            mTextInputLayout.setError(getString(R.string.passwords_do_not_match));
        }

        @Override
        public void hideError() {
            mTextInputLayout.setError(null);
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                NavUtils.navigateUpTo(this, NavUtils.getParentActivityIntent(this));
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        return true;
    }
}

File: app/src/main/java/com/softartdev/notecrypt/ui/settings/security/SecurityActivity.kt
package com.softartdev.notecrypt.ui.settings.security

import android.annotation.SuppressLint
import android.os.Bundle
import android.support.design.widget.TextInputLayout
import android.support.v4.app.NavUtils
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatDelegate
import android.view.Menu
import android.view.MenuItem
import android.widget.CompoundButton
import android.widget.EditText
import com.softartdev.notecrypt.R
import com.softartdev.notecrypt.ui.base.BaseActivity
import kotlinx.android.synthetic.main.activity_security.*
import kotlinx.android.synthetic.main.dialog_change_password.view.*
import kotlinx.android.synthetic.main.dialog_password.view.*
import kotlinx.android.synthetic.main.dialog_set_password.view.*
import javax.inject.Inject

class SecurityActivity : BaseActivity(), SecurityView, CompoundButton.OnCheckedChangeListener {

    companion object {
        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }
    }

    @Inject lateinit var mPresenter: SecurityPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_security)
        activityComponent().inject(this)
        mPresenter.attachView(this)

        enable_encryption_switch.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_lock_black_24dp, 0, 0, 0)
        enable_encryption_switch.isChecked = mPresenter.isEncryption
        enable_encryption_switch.setOnCheckedChangeListener(this)

        set_password_button.setCompoundDrawablesWithIntrinsicBounds(R.drawable.ic_password_black_24dp, 0, 0, 0)
        set_password_button.setOnClickListener { mPresenter.changePassword() }
    }

    override fun showEncryptEnable(encryption: Boolean) {
        enable_encryption_switch.setOnCheckedChangeListener(null)
        enable_encryption_switch.isChecked = encryption
        enable_encryption_switch.setOnCheckedChangeListener(this)
    }

    override fun onCheckedChanged(buttonView: CompoundButton?, isChecked: Boolean) {
        mPresenter.changeEncryption(isChecked)
    }

    override fun showPasswordDialog() {
        @SuppressLint("InflateParams") val dialogPasswordView = layoutInflater.inflate(R.layout.dialog_password, null)
        val alertDialog = with(AlertDialog.Builder(this)) {
            setView(dialogPasswordView)
            setPositiveButton(android.R.string.ok, null)
            setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }
            create()
        }
        alertDialog.setOnShowListener {
            val okButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE)
            okButton.setOnClickListener {
                val pass = PassMediator(dialogPasswordView.enter_password_text_input_layout, dialogPasswordView.enter_password_edit_text)
                if (mPresenter.enterPassCorrect(pass)) {
                    alertDialog.dismiss()
                }
            }
        }
        alertDialog.show()
    }

    override fun showSetPasswordDialog() {
        @SuppressLint("InflateParams") val dialogPasswordView = layoutInflater.inflate(R.layout.dialog_set_password, null)
        val alertDialog = with(AlertDialog.Builder(this)) {
            setView(dialogPasswordView)
            setPositiveButton(android.R.string.ok, null)
            setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }
            create()
        }
        alertDialog.setOnShowListener {
            val okButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE)
            okButton.setOnClickListener {
                val pass = PassMediator(dialogPasswordView.set_password_text_input_layout, dialogPasswordView.set_password_edit_text)
                val repeatPass = PassMediator(dialogPasswordView.repeat_set_password_text_input_layout, dialogPasswordView.repeat_set_password_edit_text)
                if (mPresenter.setPassCorrect(pass, repeatPass)) {
                    alertDialog.dismiss()
                }
            }
        }
        alertDialog.show()
    }

    override fun showChangePasswordDialog() {
        @SuppressLint("InflateParams") val dialogPasswordView = layoutInflater.inflate(R.layout.dialog_change_password, null)
        val alertDialog = with(AlertDialog.Builder(this)) {
            setView(dialogPasswordView)
            setPositiveButton(android.R.string.ok, null)
            setNegativeButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }
            create()
        }
        alertDialog.setOnShowListener {
            val okButton = alertDialog.getButton(AlertDialog.BUTTON_POSITIVE)
            okButton.setOnClickListener {
                val oldPass = PassMediator(dialogPasswordView.old_password_text_input_layout, dialogPasswordView.old_password_edit_text)
                val newPass = PassMediator(dialogPasswordView.new_password_text_input_layout, dialogPasswordView.new_password_edit_text)
                val repeatNewPass = PassMediator(dialogPasswordView.repeat_new_password_text_input_layout, dialogPasswordView.repeat_new_password_edit_text)
                if (mPresenter.changePassCorrect(oldPass, newPass, repeatNewPass)) {
                    alertDialog.dismiss()
                }
            }
        }
        alertDialog.show()
    }

    private inner class PassMediator internal constructor(internal var mTextInputLayout: TextInputLayout, internal var mEditText: EditText) : SecurityView.DialogDirector {

        override val textString: String?
            get() = mEditText.text.toString()

        override fun showIncorrectPasswordError() {
            mTextInputLayout.error = getString(R.string.incorrect_password)
        }

        override fun showEmptyPasswordError() {
            mTextInputLayout.error = getString(R.string.empty_password)
        }

        override fun showPasswordsNoMatchError() {
            mTextInputLayout.error = getString(R.string.passwords_do_not_match)
        }

        override fun hideError() {
            mTextInputLayout.error = null
        }
    }

    override fun showError(message: String?) {
        with(AlertDialog.Builder(this)) {
            setTitle(android.R.string.dialog_alert_title)
            setMessage(message)
            setNeutralButton(android.R.string.cancel) { dialog, _ -> dialog.cancel() }
            show()
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                NavUtils.getParentActivityIntent(this)?.let { NavUtils.navigateUpTo(this@SecurityActivity, it) }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        return true
    }
}
